src/platform/datapath_winuser.c

```diff
@@ -4080,198 +4034,65 @@ CxPlatSocketSend(
+            Cqe->lpOverlapped,
+            CXPLAT_SEND_DATA,
+            Sqe);
 
 #ifdef CXPLAT_DATAPATH_QUEUE_SENDS
-            if (NumberOfBytesTransferred == UINT32_MAX &&
-                CxPlatRundownAcquire(&SocketProc->UpcallRundown)) {
-                CxPlatSocketSendInline(
-                    SocketProc,
-                    &SendData->LocalAddress,
-                    &SendData->RemoteAddress,
-                    SendData);
-                CxPlatRundownRelease(&SocketProc->UpcallRundown);
-            } else
+    if (Cqe->dwNumberOfBytesTransferred == UINT32_MAX &&
+        CxPlatRundownAcquire(&SocketProc->UpcallRundown)) {
+        CxPlatSocketSendInline(
+            SocketProc,
+            &SendData->LocalAddress,
+            &SendData->RemoteAddress,
+            SendData);
+        CxPlatRundownRelease(&SocketProc->UpcallRundown);
+    } else
 #endif // CXPLAT_DATAPATH_QUEUE_SENDS
-            {
-                CxPlatSendDataComplete(
-                    SocketProc,
-                    SendData,
-                    IoResult);
-            }
-        }
+    {
+        CxPlatSendDataComplete(
+            SocketProc,
+            SendData,
+            RtlNtStatusToDosError((NTSTATUS)Cqe->Internal));
     }
-
-    return TRUE;
-}
-
-_IRQL_requires_max_(PASSIVE_LEVEL)
-QUIC_STATUS
-CxPlatSocketSetParam(
-    _In_ CXPLAT_SOCKET* Socket,
-    _In_ uint32_t Param,
-    _In_ uint32_t BufferLength,
-    _In_reads_bytes_(BufferLength) const UINT8 * Buffer
-    )
-{
-    UNREFERENCED_PARAMETER(Socket);
-    UNREFERENCED_PARAMETER(Param);
-    UNREFERENCED_PARAMETER(BufferLength);
-    UNREFERENCED_PARAMETER(Buffer);
-    return QUIC_STATUS_NOT_SUPPORTED;
 }
 
-_IRQL_requires_max_(PASSIVE_LEVEL)
-QUIC_STATUS
-CxPlatSocketGetParam(
-    _In_ CXPLAT_SOCKET* Socket,
-    _In_ uint32_t Param,
-    _Inout_ PUINT32 BufferLength,
-    _Out_writes_bytes_opt_(*BufferLength) UINT8 * Buffer
+void
+CxPlatDataPathProcessCqe(
+    _In_ CXPLAT_CQE* Cqe
     )
 {
-    UNREFERENCED_PARAMETER(Socket);
-    UNREFERENCED_PARAMETER(Param);
-    UNREFERENCED_PARAMETER(BufferLength);
-    UNREFERENCED_PARAMETER(Buffer);
-    return QUIC_STATUS_NOT_SUPPORTED;
+    switch (CxPlatCqeType(Cqe)) {
+    case CXPLAT_CQE_TYPE_DATAPATH_SHUTDOWN: {
+        CXPLAT_DATAPATH_PROC* DatapathProc =
+            CONTAINING_RECORD(CxPlatCqeUserData(Cqe), CXPLAT_DATAPATH_PROC, ShutdownSqe);
+        CxPlatProcessorContextUninitializeComplete(DatapathProc);
+        break;
+    }
+    case CXPLAT_CQE_TYPE_SOCKET_SHUTDOWN: {
+        CXPLAT_SOCKET_PROC* SocketProc =
+            CONTAINING_RECORD(CxPlatCqeUserData(Cqe), CXPLAT_SOCKET_PROC, ShutdownSqe);
+        CxPlatSocketContextUninitializeComplete(SocketProc);
+        break;
+    }
+    case CXPLAT_CQE_TYPE_SOCKET_IO: {
+        CXPLAT_SOCKET_PROC* SocketProc =
+            CONTAINING_RECORD(CxPlatCqeUserData(Cqe), CXPLAT_SOCKET_PROC, IoSqe);
+        if (Cqe->lpOverlapped == &SocketProc->IoSqe.Sqe) {
```

Comment:
I don't understand this check. Should "IoSqe" have "recv" in the name or something?

Reply:
The only way to know if it's a send completion or a receive completion is by the OVERLAPPED returned. Receive completions all point to the socket processor's OVERLAPPED. Send completions are per-send. So if it's not `&SocketProc->IoSqe.Sqe` it must be send.

As to the name, the "user data" for both still returns `&SocketProc->IoSqe`, so it's general because of that.

The intricacies of the various network IO models is insane...

Reply:
......k